### PR TITLE
library: open module relative to libospray's location

### PR DESCRIPTION
**Pull request by mathstuf**
See #311

### CODE REVIEW

**Comment by mathstuf** [comment]
Cc: @demarle

**Comment by demarle** [comment]
I give it a thumbs up.

**Comment by jeffamstutz** [comment]
Thanks for contributing this! I'll work on a `v1.8.4` release today and get this out ASAP (if everything checks out). FWIW, this is already on the `release-1.8.x` branch if you'd like to consume it on your end immediately.

**Comment by jeffamstutz** [comment]
Can you confirm that this does actually fix the issue you were seeing before I cut the `v1.8.4` release?

**Comment by mathstuf** [comment]
With the patch applied, our macOS binaries are much happier now (also worked on my Linux machine). Windows was untested.

File: components/ospcommon/library.cpp [hunk 1]
```diff
@@ -29,6 +29,50 @@
 #include <sys/times.h>
 #endif
 
+extern "C" {
+
+  /* Export a symbol to ask the dynamic loader about in order to locate this
+   * library at runtime. */
+  OSPCOMMON_INTERFACE const int _ospray_anchor()
```
Comment (mathstuf) [inline components/ospcommon/library.cpp]: The `const` here is leftover from previous iterations on this function. It should be removed. Want a PR for it or is it easier for you to just directly commit the change?

Comment (jeffamstutz) [reply components/ospcommon/library.cpp]: Already removed it locally in ‘release-1.8.x’. :)

**Comment by jeffamstutz** [comment]
Good enough for me, thanks!

**Comment by johguenther** [comment]
Thanks for the quick workaround.
I still like to put up some thoughts for discussion:
- if an additional/specific path to search libraries is used, that should come as "last resort", because otherwise users cannot override this (e.g. with `LD_PRELOAD`, or maybe also with `lmod` modules) &ndash;  there is actually a big comment in `library.cpp`:
    ```
    // do NOT try to find the library in another location
    // if you want that use LD_LIBRARY_PATH or equivalents
    ```
- the used path is actually the one from `libospray_common.so`, not from `libospray.so`; we think about splitting out "ospcommon" as a separate library, which then may be used by multiple projects to load modules; thus in future we need to generalize that path 
- why not use `RPATH` (at least on Linux and OSX)?

**Comment by mathstuf** [comment]
> if an additional/specific path to search libraries is used, that should come as "last resort", because otherwise users cannot override this (e.g. with LD_PRELOAD, or maybe also with lmod modules)

What is the use case for custom modules? What are the facilities for providing ones own OSPRay backend plugin?

> why not use RPATH (at least on Linux and OSX)?

This would work today with an rpath (or runpath) of `$ORIGIN`. I'm not sure how this would generalize with `libospcommon` and `libospray` being in different prefixes and `libospray` wanting its plugins to be found relative to it (`DT_RUNPATH` is the New Way™ and is not inherited like `DT_RPATH` is nevermind what would happen if two libraries in different prefixes end up loading `libospcommon` since `ld.so` would elide the second loading).

The VTK way where a symbol name (Windows) or symbol pointer (`dladdr`) is given to use as the lookup key would work because then the caller of `libospcommon` could then say "look up relative to the library providing X" where X is something like the `_ospray_anchor` function I added.

**Comment by johguenther** [comment]
Thanks for the feedback. In a future release we like `libospcommon` to be a bit more independent of OSPRay, i.e., not prefixing a path in `library.cpp`. Instead, this functionality could move into  `ospLoadModule`, i.e., OSPRay will provide a path to look for plugins, like you suggested.

**Pull request merged**